griptape/artifacts/list_artifact.py

```diff
@@ -1,23 +1,25 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Optional
+from typing import TYPE_CHECKING, Generic, Optional, TypeVar
 
 from attrs import Attribute, define, field
 
 from griptape.artifacts import BaseArtifact
 
 if TYPE_CHECKING:
-    from collections.abc import Sequence
+    from collections.abc import Iterator, Sequence
+
+T = TypeVar("T", bound=BaseArtifact)
 
 
 @define
-class ListArtifact(BaseArtifact):
-    value: Sequence[BaseArtifact] = field(factory=list, metadata={"serializable": True})
+class ListArtifact(BaseArtifact, Generic[T]):
+    value: Sequence[T] = field(factory=list, metadata={"serializable": True})
```

Comment:
`list` is not covariant, `Sequence` is. Here is an example pyright error:

```
    Return type mismatch: base method returns type "BaseArtifact | list[BaseArtifact]", override returns type "list[CsvRowArtifact]"
      Type "list[CsvRowArtifact]" is incompatible with type "BaseArtifact | list[BaseArtifact]"
        "list[CsvRowArtifact]" is incompatible with "BaseArtifact"
        "list[CsvRowArtifact]" is incompatible with "list[BaseArtifact]"
          Type parameter "_T@list" is invariant, but "CsvRowArtifact" is not the same as "BaseArtifact"
          Consider switching from "list" to "Sequence" which is covariant (reportIncompatibleMethodOverride)
```

Reply:
TIL!

After looking into why `list` is not covariant and why `Sequence` is, its because `list` is mutable!

example:
```py
class Animal:
    pass

class Dog(Animal):
    pass

dogs: List[Dog] = [Dog(), Dog()]
animals: List[Animal] = dogs  # This would be allowed if lists were covariant

animals.append(Animal())  # Adding an Animal (not necessarily a Dog!) to a list of Dogs
```

So this leads me to the question, `ListArtifact` is not intended to be mutable, is it?

Reply:
> ListArtifact is not intended to be mutable, is it?

Good question, probably not. I can investigate changing back to `list`, but it will have implications throughout the codebase so I think it should be done in a separate PR.